tools/hsmtool.c

```diff
@@ -368,6 +371,55 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
 	return 1;
 }
 
+static int dumponchaindescriptors(const char *hsm_secret_path, const char *passwd)
+{
+	struct secret hsm_secret;
+	u8 bip32_seed[BIP32_ENTROPY_LEN_256];
+	u32 salt = 0;
+	struct ext_key master_extkey;
+	char *enc_xpub, *descriptor, checksum[DESCRIPTOR_CHECKSUM_LENGTH + 1];
+
+	if (passwd)
+		get_encrypted_hsm_secret(&hsm_secret, hsm_secret_path, passwd);
+	else
+		get_hsm_secret(&hsm_secret, hsm_secret_path);
+
+	/* We use m/0/0/k as the derivation tree for onchain funds. */
+
+	/* The root seed is derived from hsm_secret using hkdf.. */
+	do {
+		hkdf_sha256(bip32_seed, sizeof(bip32_seed),
+			    &salt, sizeof(salt),
+			    &hsm_secret, sizeof(hsm_secret),
+			    "bip32 seed", strlen("bip32 seed"));
+		salt++;
+		/* ..Which is used to derive m/ */
+	} while (bip32_key_from_seed(bip32_seed, sizeof(bip32_seed),
+				     /* An xpub can easily be converted to a tpub */
```

Comment:
~No. This has potential to be *extremely* confusing. We need to either be told what chain to output it for or figure it out from the directory structure.~

updated later.